### ISSUE TITLE
[A WISH] Create a "reconciliation mode" for hledger-ui

### ISSUE THREAD

**Issue by the-solipsist** [simonmichael/hledger#632]
While generally it is a great principle that the tools like `hledger` should not change the ledger file they are dealing with, given the existence of hledger-add and hledger-web, both of which allow for adding of transactions, I feel there could be a tiny bit of text editing that should be allowable in hledger-ui as well.

One frequent task for any accounting practice is to reconcile one's local statement of accounts with the bank's statements.  

hledger allows for three statuses — unmarked, pending, or cleared — that can be used to indicate reconciliation status, for instance.  It would be very useful to change the status of individual transactions from within the hledger-ui without having to exit the programme (i.e., via `E`) for each transaction, or having to do this in a text editor.

**Comment by simonmichael** [comment]
I agree, and also want this feature. It requires some new infrastructure: support for rewriting an arbitrary transaction in the middle of a file. An API for this needs to be drafted, clarifying how we specify the transaction (eg by file and line number and/or transaction index).

**Comment by adept** [comment]
Idea : Reconciliation mode could just append the entries at the end of journal or even to separate file. They would be auto transactions that declare change of status and refer to original transactions by their hash. This allows changing /re reading journal on the fly. 

Then "hledger rewrite" could be used to generate updated journal in one pass or change mode can do this when you somehow indicate that you are done. This could be implemented as a change to journal reader which will hash each transaction after parsing, and if hash is mentioned in the last of rewrites, it will remember current file and position and return positions as result of parsing. Separate pass then changes all the files, resulting in least invasive changes.

**Comment by simonmichael** [comment]
Related:

- https://hledger.org/scripts.html#hledger-edit edits individual entries somehow (using their file position info I assume). [So does [hledger-textual](https://github.com/thesmokinator/hledger-textual)]

- The `tindex` (transaction index) field, which can be seen in CSV/JSON output, uniquely numbers each transaction across all files, for a particular read of those files, but it's not permanent - if files are reloaded (eg due to a change in one of them), all transactions are renumbered, possibly changing tindex values.

- We can't calculate unique ids just by hashing a transaction's data, because identical records are fairly common in bank data.

- We could calculate a unique id at `import` time, assuming the usual [import requirements](https://hledger.org/hledger.html#overlap-detection), based on (eg): the account or csv file, the date, and the intra-day sequence number (eg the second txn seen on this date in this csv file).

- Saving such ids permanently probably requires adding them to each entry as a visible tag, even though this adds noise for human editors.

- Or we could imagine saving them somewhere off to the side, like an sqlite db, along with file position info. But that sounds fragile - edits to the file could easily break the correspondence.

**Comment by jack9603301** [comment]
> Or we could imagine saving them somewhere off to the side, like an sqlite db, along with file position info. But that sounds fragile - edits to the file could easily break the correspondence.

Why not use an in-memory database?

> We can't calculate unique ids just by hashing a transaction's data, because identical records are fairly common in bank data.

You can use UUID or GUID encoding
